### PR TITLE
Enable composable benchmark configs for flexible model+device+optimization scheduling

### PR DESCRIPTION
**Pull request by guangy10**
It's not true that any combination of model + delegate can work, which make adding a new model to continuous run not easy as the workflow will run it across all delegates. Besides, each delegate may run with different configurations. For example, llama3.2 spinquant is using a prequantized checkpoitn hence it's not using the recipe for the regular fp32 checkpint. To support various combinations and optimizations, we are migrating to use `benchmark_configs` which is a set of predefined configs with combination of all optimizations that are possibly applied to the model, e.g. kv cache, embedding/activation quant, dtype, delegation, sdpa, etc.

In this PR, given a model (either a Hugging Face model ID or a in-tree model name) and a target platform ("android" vs "ios"), it's retrieving a list of supported benchmark configurations from the script `gather_benchmark_configs.py` and schedule the benchmark jobs accordingly. From the workflow dispatcher, users will just need to enter the model names, it will discover all supported benchmark configs for each model. Further more (not included in this PR), we can potentially expose `config_args` (key-value paris) from the script, if there is a way to store them in the DB and display in the dashboard. It will help understand how exactly a model is exported/lowered when discussing/debugging perf metrics.

Apple: https://github.com/pytorch/executorch/actions/runs/12404655922
Android: https://github.com/pytorch/executorch/actions/runs/12404651330

### CODE REVIEW

**Comment by guangy10** [comment]
All apple jobs are failing at "Setting up CI environment" step. The log doesn't have much useful info to tell if it's relevant. Scheduled one workflow to compare: https://github.com/pytorch/executorch/actions/runs/12387135036

File: .github/workflows/apple-perf.yml
Comment (huydhn) [inline .github/workflows/apple-perf.yml:156]: This line and the one on line 161 looks wrong, you probably mean `${{ matrix.config }}` here.  The failure looks cryptic, but it tells something wrong in the script https://github.com/pytorch/executorch/actions/runs/12387012961/job/34575783485#step:9:8209

Comment (huydhn) [inline .github/workflows/apple-perf.yml:156]:
```suggestion
        if [[ ${{ matrix.config }} == *"coreml"* ]]; then
```

Comment (guangy10) [reply .github/workflows/apple-perf.yml:156]: Oops, forget to fix this

**Comment by guangy10** [comment]
```
Dec 18 09:37:37 PDX647397983 kernel[0] <Notice>: memorystatus: killing process 602 [Benchmark] in high band FOREGROUND (100) - memorystatus_available_pages: 112541
Dec 18 09:37:37 PDX647397983 kernel[0] <Notice>: Benchmark[602] Corpse allowed 1 of 5
Dec 18 09:37:37 PDX647397983 backboardd(IOKit)[65] <Notice>: Connection removed: IOHIDEventSystemConnection uuid:2A3BA7CC-8C4A-496C-ACAA-32E7D50DC9B5 pid:602 process:Benchmark type:Passive entitlements:0x0 caller:BackBoardServices: <redacted> + 280 attributes:{
    HighFrequency = 1;
    bundleID = "org.pytorch.executorch.Benchmark";
    pid = 602;
} state:0x1 events:0 mask:0x0 dropped:0 dropStatus:0 droppedMask:0x0 lastDroppedTime:NONE
Dec 18 09:37:37 PDX647397983 kernel[0] <Notice>: 7347.894 memorystatus: killing_specific_process pid 602 [Benchmark] (per-process-limit 100) 3145733KB - memorystatus_available_pages: 112815
```

It looks like **Llama3.2 1B w/ fb16** is running around the 3GB RAM limits for iOS set by AWS.
  - OOMed: https://github.com/pytorch/executorch/actions/runs/12388301115/job/34607987766
  - Passed: https://github.com/pytorch/executorch/actions/runs/12386167313/job/34574116322
  
We will probably not include this path in the continuous run until the RAM limits is lifted.

**Comment by guangy10** [comment]
The failure in [apple-perf / benchmark-on-device (llama, mps, arn:aws:devicefarm:us-west-2:308535385114:devicepool:02a2cf0f-6d... / mobile-job (ios)](https://hud.pytorch.org/pr/pytorch/executorch/7349#34614070197) ([gh](https://github.com/pytorch/executorch/actions/runs/12399052782/job/34614070197)) is pre-existing. See details in https://github.com/pytorch/executorch/issues/7366

**Comment by guangy10** [comment]
The android failure [android-perf / benchmark-on-device (llama, xnnpack_q8, arn:aws:devicefarm:us-west-2:308535385114:devicepool:02a2... / mobile-job (android)](https://hud.pytorch.org/pr/pytorch/executorch/7349#34616060368) ([gh](https://github.com/pytorch/executorch/actions/runs/12399052791/job/34616060368)) seem irrelevant to this PR since the [same model+config passed](https://github.com/pytorch/executorch/actions/runs/12386663978) earlier on this PR. But it's still important to figure out the root cause as it doesn't seem to be transient (retried and failed). @huydhn if you have any insights

**Comment by huydhn** [comment]
> The android failure [android-perf / benchmark-on-device (llama, xnnpack_q8, arn:aws:devicefarm:us-west-2:308535385114:devicepool:02a2... / mobile-job (android)](https://hud.pytorch.org/pr/pytorch/executorch/7349#34616060368) ([gh](https://github.com/pytorch/executorch/actions/runs/12399052791/job/34616060368)) seem irrelevant to this PR since the [same model+config passed](https://github.com/pytorch/executorch/actions/runs/12386663978) earlier on this PR. But it's still important to figure out the root cause as it doesn't seem to be transient (retried and failed). @huydhn if you have any insights

I think the script returns the wrong device pool here for Android job.  From https://github.com/pytorch/executorch/actions/runs/12401286750/job/34620341884?pr=7349#step:4:35, it returned `arn:aws:devicefarm:us-west-2:308535385114:devicepool:02a2cf0f-6d9b-45ee-ba1a-a086587469e6/3b5acd2e-92e2-4778-b651-7726bafe129d` which is the iPhone device pool

File: .github/workflows/android-perf.yml
Comment (guangy10) [reply .github/workflows/android-perf.yml:74]: What did I do!

**Comment by guangy10** [comment]
> > The android failure [android-perf / benchmark-on-device (llama, xnnpack_q8, arn:aws:devicefarm:us-west-2:308535385114:devicepool:02a2... / mobile-job (android)](https://hud.pytorch.org/pr/pytorch/executorch/7349#34616060368) ([gh](https://github.com/pytorch/executorch/actions/runs/12399052791/job/34616060368)) seem irrelevant to this PR since the [same model+config passed](https://github.com/pytorch/executorch/actions/runs/12386663978) earlier on this PR. But it's still important to figure out the root cause as it doesn't seem to be transient (retried and failed). @huydhn if you have any insights
> 
> I think the script returns the wrong device pool here for Android job. From https://github.com/pytorch/executorch/actions/runs/12401286750/job/34620341884?pr=7349#step:4:35, it returned `arn:aws:devicefarm:us-west-2:308535385114:devicepool:02a2cf0f-6d9b-45ee-ba1a-a086587469e6/3b5acd2e-92e2-4778-b651-7726bafe129d` which is the iPhone device pool

Aha! I should probably return the device pool name as well, so that we can easily spot if it's using the wrong device pool

**Comment by huydhn** [comment]
> We will probably not include this path in the continuous run until the RAM limits is lifted.

I'm still following up with AWS on this.  The confusing part is I'm not entirely sure if this is something from AWS (but let's see).  From what I read in https://developer.apple.com/documentation/bundleresources/entitlements/com.apple.developer.kernel.increased-memory-limit, this feels like a maybe and not a guarantee

>  If you use this entitlement, make sure your app still behaves correctly if additional memory isn’t available.

and 

> An increased memory limit is only available on some device models, call the os_proc_available_memory function to determine the amount of memory available

File: .github/workflows/android-perf.yml
Comment (guangy10) [inline .github/workflows/android-perf.yml:132]: @huydhn do you know how I can customize the job title? It looks like it's automatically generated as part of the `matrix:` configuration. Currently it's displayed as "[export-models (meta-llama/Llama-3.2-1B, llama3_htp, arn:aws:devicefarm:us-west-2:308535385114:dev... / linux-job](https://github.com/pytorch/executorch/actions/runs/12401578767/job/34621284740#logs)", where the arn is not useful, I'd like to replace it with the human readable device pool name.

**Comment by guangy10** [comment]
Core ML ANE job: https://github.com/pytorch/executorch/actions/runs/12402205998/job/34624007233

**Comment by guangy10** [comment]
Qualcomm HTP job: https://github.com/pytorch/executorch/actions/runs/12402906809

File: .ci/scripts/gather_benchmark_configs.py
Comment (huydhn) [inline .ci/scripts/gather_benchmark_configs.py:137]: What is the separator you are using to split the model name and the benchmark config?  I think we should use something unique , i.e. `__` (double underscore) or something similar so that the extract script can set them correctly later

File: .github/workflows/android-perf.yml
Comment (huydhn) [reply .github/workflows/android-perf.yml:132]: AFAIK, there is no way to change this behavior from GitHub.  The name always include all the matrix dimensions.  Maybe there is a way to create a more human-friendly alias for the pool on AWS, let me take a look from that angle

**Review by huydhn — APPROVED** [review]

File: .ci/scripts/gather_benchmark_configs.py
Comment (guangy10) [reply .ci/scripts/gather_benchmark_configs.py:137]: Can we use some more reliable than the file name? For example, can we store the benchmark config as it's serializable to json, then it's model name, benchmark_config (delegate, quant, dtype, etc.) can be safely retrieved.

File: .github/workflows/android-perf.yml
Comment (guangy10) [reply .github/workflows/android-perf.yml:132]: Don't worry. No need to waste time on this. I can always include the device_name in the matrix

File: .ci/scripts/gather_benchmark_configs.py
Comment (huydhn) [reply .ci/scripts/gather_benchmark_configs.py:137]: Yeah, being able to get away with parsing the filename is even better.  Let's have the benchmark config uploaded together with the export model and I think we can pick it up from there.

File: .github/workflows/android-perf.yml
Comment (guangy10) [reply .github/workflows/android-perf.yml:132]: Added "device_name" and "device_arn" to the matrix, and put the "device_arn" at the end. Now we can see each job is titled with device pool name

**Comment by guangy10** [comment]
@huydhn FYI, Core ML ANE and Qualcomm HTP jobs are not going to block this PR. I will merge this PR tonight to unblock you, and leave adding those paths separately if they are not working out of the box

**Comment by huydhn** [comment]
> @huydhn FYI, Core ML ANE and Qualcomm HTP jobs are not going to block this PR. I will merge this PR tonight to unblock you, and leave adding those paths separately if they are not working out of the box

Sounds good!  Once this lands, I could start working on bring the benchmark config to the dashboard

**Comment by guangy10** [comment]
Core ML ANE and Qualcomm HTP are not ready. Will enable in separate PRs. For now, we just hide these paths from `benchmark_configs`.

**Comment by guangy10** [comment]
@kirklandsign FYI, the results of newly added llama3 benchmarkings (spintquant, qlora, bf16) are also missing from the benchmark_results.json. See the upload job here: https://github.com/pytorch/executorch/actions/runs/12404651330/job/34631545501

**Pull request merged**
